docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,194 @@
+
+The skeleton might look something like this::
+
+    def algorithm():
+        # setup
+        v = ...
+        while True:
+            # take a step
+            d = scipy.linalg.norm(v)
+            if d < tolerance:
+                break
+
+Now, let's further assume, that you have heard of Numba and you now wish to use
+it to accelerate your function. However, after adding the
+``jit(nopython=True)``
+decorator, Numba complains that ``scipy.linalg.norm`` isn't supported. From
+looking at the documentation, you realize that a norm is probably fairly easy
+to implement using NumPy and you wanted to learn a little about the inner
+workings of Numba anyway. So, after some back and forth with the
+core-developers via Gitter you receive the following annotated template in your
+INBOX:
+
+.. literalinclude:: template.py
+
+After some deliberation and tinkering, you end up with the following code:
+
+.. literalinclude:: mynorm.py
+
+As you can see, the implementation only supports what you need right now:
+
+* Only supports 
+* Only integer and floating-point types
+* All vector norms
+* Only the Frobenius norm for matrices
+* Code sharing between vector and matrix implementations using
+  ``@register_jitable``.
+* Norms are implemented using NumPy syntax. (This is possible because
+  Numba is very aware of NumPy and many functions are supported.)
+
+So what actually happens here? The ``overload`` decorator registers a suitable
+implementation for ``scipy.linalg.norm`` in case a call to this is encountered
+in code that is being JIT-compiled, for example when you decorate your
+``algorithm`` function with ``@jit(nopython=True)``. In that case, the function
+``jit_norm`` will be called with the currently encountered types and will then
+return either ``_oneD_norm_x`` in the vector case and ``_two_D_norm_2``.
+
+You can download the example code here: :download:`mynorm.py </extending/mynorm.py>`
+
+Implementing ``@overload`` for NumPy functions
+==============================================
+
+Numba supports NumPy through the provision of ``@jit`` compatible
+re-implementations of NumPy functions. In such cases ``@overload`` is a very
+convenient option for writing such implementations, however there are a few
+additional things to watch out for.
+
+* The Numba implementation should match the NumPy implementation as closely as
+  feasible with respect to accepted types, arguments, raised exceptions and
+  algorithmic complexity (Big-O / Landau order).
+
+* When implementing supported argument types, bear in mind that, due to
+  duck typing, NumPy does tend to accept a multitude of argument types beyond
+  NumPy arrays such as scalar, list, tuple, set, iterator, generator etc.
+  You will need to account for that during type inference and subsequently as
+  part of the tests.
+
+* A NumPy function may return a scalar, array or a data structure
+  which matches one of its inputs, you need to watch be aware of type
```

Comment:
`you need to watch be aware of type` <- needs fixing.

Reply:
yes